drivers/flash/flash_ambiq.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2023 Ambiq Micro Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ambiq_flash_controller
+
+#include <zephyr/kernel.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/logging/log.h>
+
+#include <am_mcu_apollo.h>
+
+LOG_MODULE_REGISTER(flash_ambiq, CONFIG_FLASH_LOG_LEVEL);
+
+#define SOC_NV_FLASH_NODE      DT_INST(0, soc_nv_flash)
+#define SOC_NV_FLASH_ADDR      DT_REG_ADDR(SOC_NV_FLASH_NODE)
+#define SOC_NV_FLASH_SIZE      DT_REG_SIZE(SOC_NV_FLASH_NODE)
+#define FLASH_WRITE_BLOCK_SIZE DT_PROP(SOC_NV_FLASH_NODE, write_block_size)
+#define FLASH_ERASE_BLOCK_SIZE DT_PROP(SOC_NV_FLASH_NODE, erase_block_size)
+
+#if defined(CONFIG_MULTITHREADING)
+static struct k_sem flash_ambiq_sem;
+#define FLASH_SEM_INIT() k_sem_init(&flash_ambiq_sem, 1, 1)
+#define FLASH_SEM_TAKE() k_sem_take(&flash_ambiq_sem, K_FOREVER)
+#define FLASH_SEM_GIVE() k_sem_give(&flash_ambiq_sem)
+#else
+#define FLASH_SEM_INIT()
+#define FLASH_SEM_TAKE()
+#define FLASH_SEM_GIVE()
+#endif /* CONFIG_MULTITHREADING */
+
+static const struct flash_parameters flash_ambiq_parameters = {
+	.write_block_size = FLASH_WRITE_BLOCK_SIZE,
```

Comment:
I think assignments like this should be `MAX(hardware_specific, DTS value)`. And there probably should be build error when DTS is smaller than hardware minimal and not a multiply of it?

Reply:
I don't clearly understand this point. The write_block_size parameter here is set by the DT_PROP(SOC_NV_FLASH_NODE, write_block_size) which should be configured properly in DTS. Are you saying we need to define a macro for the hardware_specific requirement (16 for Ambiq Apollo4x)?

#define HW_WRITE_BLOCK_SIZE  16
.write_block_size = MAX(HW_WRITE_BLOCK_SIZE, FLASH_WRITE_BLOCK_SIZE)

Reply:
I also don't really understand this point. The write-block-size parameter is a real hardware device property and the value in the dts should be correct.

For the erase-block-size (this is a "make-compatible" value and could be adapted to fit the needs) I could understand that it is checked as being a multiple of write-block-size.

Reply:
My point is simple here: the device has design constrains on the write block size that should be maintained by the driver at compilation phase. The user may override write block size in DTS with value that still is correct and will work, like multiple of the hardware characteristic of the write-block-size.
That is why the write-block-size taken from the DTS should be checked against what the driver will actually accept or could work with.
I know that there are drivers that basically ignore the DTS and have the value encoded in driver only, so the DTS is more informative than actually used for anything, but here it has been decided to take the values as something that is used to compile the driver, so the driver should have compile time checks on the user input.